### PR TITLE
Allow `uses_mappings()` to find mappings that appear in every layer of the plot

### LINKED ISSUE #34: `uses_mappings(..., local_only = FALSE)` should check if mappings exist in all layers of plot

### PR DESCRIPTION
**Pull request by rossellhayes**
If `local_only = FALSE`, `get_mappings()` and `uses_mappings()` find mappings that appear in `ggplot()` _or_ in every layer of the plot

``` r
library(ggplot2)
library(ggcheck)

p <- ggplot(mtcars, aes(x = wt)) +
  geom_point(aes(y = mpg, group = cyl)) +
  geom_smooth(aes(y = mpg, group = am))

get_mappings(p, local_only = TRUE)
#> Aesthetic mapping: 
#> * `x` -> `wt`
get_mappings(p)
#> Aesthetic mapping: 
#> * `x` -> `wt`
#> * `y` -> `mpg`

uses_mappings(p, aes(x = wt, y = mpg), local_only = TRUE)
#> [1] FALSE
uses_mappings(p, aes(x = wt, y = mpg))
#> [1] TRUE

uses_mappings(p, aes(x = wt, y = mpg, group = cyl))
#> [1] FALSE
```

<sup>Created on 2022-03-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Closes #34

### CODE REVIEW

File: R/mappings.R [hunk 1]
```diff
@@ -72,7 +72,18 @@ get_mappings <- function(p, local_only = FALSE) {
...
+  if (local_only) {
+    return(aes_map)
+  }
+
+  layer_maps <- purrr::map(p$layers, "mapping")
+  intersection <- intersect_with_attr(layer_maps)
+
+  aes_map <- c(aes_map, intersection)
+  class(aes_map) <- "uneval"
+  aes_map
```
Comment (gadenbuie) [inline R/mappings.R:86]: This smells like something that needs more explanation. I can tell from reading this that you learned something about the structure of a ggplot object that required special handling here — mappings are stored differently in the layer than they are in the base plot? A comment here would be helpful and even better if paired with a test that covers the need for these lines. That way, new readers (like me!) can benefit from what you've learned.

File: R/utils.R [hunk 2]
```diff
@@ -340,3 +340,24 @@ flatten_dots <- function(...) {
...
   args
 }
+
+intersect_with_attr <- function(.l, ...) {
+  list <- c(.l, list(...))
+  attr_list <- purrr::map(list, attr_list)
+
+  intersect <- purrr::reduce(attr_list, intersect)
+  indices <- match(intersect, attr_list[[1]])
+  list[[1]][indices]
```
Comment (gadenbuie) [inline R/utils.R:350]: This also has a bit of a smell: it's probably the constant `1`, which makes me think this function is doing something _very_ specific. OTOH `intersect_with_attr()` sounds generic enough that I think I understand what such a function would do, but I wouldn't expect these lines.

File: R/utils.R [hunk 3]
```diff
@@ -340,3 +340,24 @@ flatten_dots <- function(...) {
...
+  strip_attr <- function(x) {
+    attributes(x) <- NULL
+    x
+  }
+
+  list <- purrr::pmap(
+    c(list(strip_attr(x)), attributes(x)),
+    list
+  )
+}
```
Comment (gadenbuie) [inline R/utils.R:363]:
```suggestion
merge_attr_list <- function(x) {
  x_attr <- attributes(x)
  attributes(x) <- NULL
  # vector attributes are merged element-wise into the original list
  # scalar attributes are added to every element
  purrr::pmap(c(list(x), x_attr), list)
}
```

The `pmap(x, list)` idea is very clever as a way to merge a vector attribute into each list item. This works great when the attributes are all either length 1 or `length(x)`, but will likely choke for other lengths — `pmap()` will throw a bad length error. Do we need to worry about that?

**Comment by rossellhayes** [comment]
@gadenbuie Refactoring in [982bbbe](https://github.com/rstudio/ggcheck/pull/35/commits/982bbbe55260a6553922c4a836f2de1716d84243) should take care of all your comments, but let me know if you notice anything that could still be improved!

File: R/mappings.R [hunk 4]
```diff
@@ -72,7 +72,27 @@ get_mappings <- function(p, local_only = FALSE) {
...
 get_mappings.ggplot <- function(p, local_only = FALSE) {
-  p$mapping
+  global_map <- p$mapping
+
+  if (local_only) {
+    return(global_map)
+  }
+
+  layer_maps <- purrr::map(p$layers, "mapping")
+  layer_names <- purrr::reduce(purrr::map(layer_maps, names), intersect)
```
Comment (gadenbuie) [inline R/mappings.R:82]: We should probably catch and return early with the global mappings when:

1. No layers
2. No mappings in the layers

Which brings up that we could use some test cases specifically around this feature.

**Pull request merged**